File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.java
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        List<Integer> res = new ArrayList<>();
        Set<Integer> nums = new HashSet<>();
        Set<Integer> seen = new HashSet<>();

        for (int num: nums1) {
            nums.add(num);
        }

        for (int num: nums2) {
            if (nums.contains(num) && !seen.contains(num)) {
                res.add(num);
                seen.add(num);
            }
        }

        int[] result = new int[res.size()];

        for (int i = 0; i < res.size(); i++) {
            result[i] = res.get(i);
        }

        return result;
    }
}

File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.py
class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        result = list()
        nums = set()
        seen = set()

        for num in nums1:
            nums.add(num)

        for num in nums2:
            if num in nums and num not in seen:
                result.append(num)
                seen.add(num)

        return result